### PR TITLE
feat(cell): shared to global store for single precision floating point elements.

### PR DESCRIPTION
**Pull request by haruhi55**
resolves https://github.com/TiledTensor/TiledCUDA/issues/127 resolve https://github.com/TiledTensor/TiledCUDA/issues/117 resolve https://github.com/TiledTensor/TiledCUDA/issues/133

This pull request addresses the following issues:

- [x] Load a row-major or column-major tile from global memory to shared memory with data type `float`.
- [x] Store a row-major or column-major tile from registers to shared memory with data type `float`.
- [x] Store a row-major or column-major tile from shared memory to global memory with data type `float`.
- [x] Use a swizzled layout for shared memory to avoid bank conflicts.

**NOTE**:

1. Since the swizzling function depends on the data type, the swizzled shared memory only supports `__half` data type for global-to-shared loads and shared-to-register loads.
1. For register-to-shared stores and shared-to-global stores, the swizzled shared memory layout only supports the `float` data type.

### CODE REVIEW

File: include/cell/copy/copy_atom.hpp [hunk 1]
```diff
@@ -60,58 +60,232 @@ struct LoadMatBase {
...
+        int row = 0, col = 0;
 #pragma unroll
-        for (int i = 0; i < kSegCols; ++i) {
+        for (int i = 0; i < kSegRows; ++i) {
+            row = lane_row + i * tl::num_rows<ThreadLayout>;
 #pragma unroll
-            for (int j = 0; j < kSegRows; ++j) {
-                dst[i * kRstride + j * kCstride] = src[j * kSegCols + i];
+            for (int j = 0; j < kSegCols; ++j) {
+                col = (lane_col + j * tl::num_cols<ThreadLayout>)*kElemPerSeg;
```
Comment (KuangjuX) [inline include/cell/copy/copy_atom.hpp:88]: typo: `col = (lane_col + j * tl::num_cols<ThreadLayout>) * kElemPerSeg;`

File: include/cell/copy/copy_atom.hpp [hunk 2]
```diff
@@ -60,58 +60,232 @@ struct LoadMatBase {
...
+        // with integer indices ranging from 0 to 255). `row` and `col` are used
+        // to calculate the index of an element within this 16x16 coordinate
+        // space.
+        int row = 0, col = 0;
+#pragma unroll
+        for (int i = 0; i < kSegRows; ++i) {
+            row = lane_row + i * tl::num_rows<ThreadLayout>;
+#pragma unroll
+            for (int j = 0; j < kSegCols; ++j) {
+                col = (lane_col + j * tl::num_cols<ThreadLayout>)*kElemPerSeg;
```
Comment (KuangjuX) [inline include/cell/copy/copy_atom.hpp:144]: typo: `col = (lane_col + j * tl::num_cols<ThreadLayout>) * kElemPerSeg;`

File: include/cell/copy/copy_atom.hpp [hunk 3]
```diff
@@ -60,58 +60,232 @@ struct LoadMatBase {
...
+        int lane_col = lane_col_id();
+
+        // A base tile has a fixed shape of 16x16 (a 16x16 2D coordinate space
+        // with integer indices ranging from 0 to 255). `row` and `col` are used
+        // to calculate the index of an element within this 16x16 coordinate
+        // space.
+        int row = 0, col = 0;
+#pragma unroll
+        for (int i = 0; i < kSegRows; ++i) {
+            row = (lane_row + i * tl::num_rows<ThreadLayout>)*kElemPerSeg;
```
Comment (KuangjuX) [inline include/cell/copy/copy_atom.hpp:197]: typo: `row = (lane_row + i * tl::num_rows<ThreadLayout>) * kElemPerSeg;`

File: include/cell/copy/copy_atom.hpp [hunk 4]
```diff
@@ -60,58 +60,232 @@ struct LoadMatBase {
...
+
+        // A base tile has a fixed shape of 16x16 (a 16x16 2D coordinate space
+        // with integer indices ranging from 0 to 255). `row` and `col` are used
+        // to calculate the index of an element within this 16x16 coordinate
+        // space.
+        int row = 0, col = 0;
 #pragma unroll
-        for (int i = 0; i < kSegCols; ++i) {
+        for (int i = 0; i < kSegRows; ++i) {
+            row = (lane_row + i * tl::num_rows<ThreadLayout>)*kElemPerSeg;
```
Comment (KuangjuX) [inline include/cell/copy/copy_atom.hpp:254]: typo: `row = (lane_row + i * tl::num_rows<ThreadLayout>) * kElemPerSeg;`

**Review by KuangjuX — CHANGES_REQUESTED** [review]

File: include/cell/copy/copy_atom.hpp (hunk 1 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: This line is automatically formatted by clang-format, but I'm not sure how to configure clang-format correctly. I chose another way to write it to bypass the problem.

File: include/cell/copy/copy_atom.hpp (hunk 2 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: fixed.

File: include/cell/copy/copy_atom.hpp (hunk 3 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: fixed.

File: include/cell/copy/copy_atom.hpp (hunk 4 above)
Comment (haruhi55) [reply include/cell/copy/copy_atom.hpp]: fixed.

**Review by KuangjuX — APPROVED** [review]
LGTM.

**Pull request merged**